api/valor_api/backend/metrics/text_generation.py

```diff
@@ -0,0 +1,526 @@
+        db=db,
+        label_source=models.Annotation,
+        filters=datum_filter,
+    ).subquery()
+
+    prediction_subquery = generate_query(
+        models.Prediction,
+        models.Annotation.datum_id.label("datum_id"),
+        models.Annotation.text.label("prediction_text"),
+        models.Annotation.context.label("prediction_context"),
+        db=db,
+        label_source=models.Prediction,
+        filters=prediction_filter,
+    ).subquery()
+
+    output = []
+    if any(
+        [metric in TEXT_COMPARISON_METRICS for metric in metrics_to_return]
+    ):
+        # get reference text to compare against from groundtruths
+        # use array_agg since there can be multiple references for a given datum_uid
+        groundtruth_subquery = (
+            generate_query(
+                models.GroundTruth.id,
+                models.Datum.id.label("datum_id"),
+                models.Datum.uid.label("datum_uid"),
+                models.Dataset.name.label("dataset_name"),
+                functions.array_agg(models.Annotation.text).label(
+                    "groundtruth_text"
+                ),
+                db=db,
+                label_source=models.GroundTruth,
+                filters=datum_filter,
+            )
+            .group_by(
+                models.GroundTruth.id,
+                models.Datum.id,
+                models.Datum.uid,
+                models.Dataset.name,
+            )
+            .subquery()
+        )
+
+        joint_subquery = (
+            select(
+                groundtruth_subquery.c.datum_uid,
+                groundtruth_subquery.c.dataset_name,
+                functions.array_agg(
+                    prediction_subquery.c.prediction_text
+                ).label("predictions"),
+                functions.array_agg(
+                    groundtruth_subquery.c.groundtruth_text
+                ).label("references"),
+            )
+            .select_from(groundtruth_subquery)
+            .join(
+                prediction_subquery,
+                groundtruth_subquery.c.datum_id
+                == prediction_subquery.c.datum_id,
+            )
+            .group_by(
+                groundtruth_subquery.c.datum_uid,
+                groundtruth_subquery.c.dataset_name,
+            )
+        )
+
+        results = db.execute(joint_subquery).all()
+        is_BLEU_enabled = "BLEU" in metrics_to_return
+        is_ROUGE_enabled = "ROUGE" in metrics_to_return
+
+        for datum_uid, dataset_name, predictions, references in results:
+            if is_BLEU_enabled:
+                bleu_metrics = _calculate_sentence_bleu(
+                    predictions=predictions,
+                    references=references,
+                    weights=metric_params.get(
+                        "weights", [0.25, 0.25, 0.25, 0.25]
+                    ),
+                )
+
+                output += [
+                    schemas.BLEUMetric(
```

Comment:
@ntlind This metric does not make any mention of the groundtruth that it was compared to. Is this because it is compared to all groundtruths associated to the datum and the bleu score reported is the max?